File: app/src/main/java/com/personal/accountantAssistant/ui/barcode/BarcodeScanActivity.java
package com.personal.accountantAssistant.ui.barcode;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.SparseArray;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.widget.EditText;

import com.google.android.gms.vision.CameraSource;
import com.google.android.gms.vision.Detector;
import com.google.android.gms.vision.barcode.Barcode;
import com.google.android.gms.vision.barcode.BarcodeDetector;
import com.personal.accountantAssistant.R;
import com.personal.accountantAssistant.utils.Constants;
import com.personal.accountantAssistant.utils.ParserUtils;
import com.personal.accountantAssistant.utils.PermissionsUtils;

import java.io.IOException;
import java.util.Objects;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

public class BarcodeScanActivity extends AppCompatActivity {

    private Context context;

    private SurfaceView cameraPreviewSurface;
    private EditText barcodeScanNumber;
    private CameraSource cameraSource;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_scan_barcode);
        Objects.requireNonNull(getSupportActionBar())
                .setTitle(R.string.barcode_scanner);

        context = BarcodeScanActivity.this;

        cameraPreviewSurface = findViewById(R.id.bar_code_scan_view);
        barcodeScanNumber = findViewById(R.id.bar_code_scan_number);
        /*final Button barcodeScanButton = findViewById(R.id.bar_code_scan_button);
        barcodeScanButton.setOnClickListener(view -> finish());*/

        createCameraSource();
    }

    public void createCameraSource() {

        final DisplayMetrics displayMetrics = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        final BarcodeDetector barcodeDetector = new BarcodeDetector.Builder(context)
                .setBarcodeFormats(Barcode.PRODUCT).build();

        cameraSource = new CameraSource.Builder(context, barcodeDetector)
                .setFacing(CameraSource.CAMERA_FACING_BACK)
                .setAutoFocusEnabled(Boolean.TRUE)
                .setRequestedPreviewSize(displayMetrics.widthPixels, displayMetrics.heightPixels)
                .setRequestedFps(24.0f)
                .build();

        cameraPreviewSurface
                .getHolder()
                .addCallback(new SurfaceHolder.Callback() {
                    @Override
                    public void surfaceCreated(SurfaceHolder surfaceHolder) {
                        cameraSourceStartFrom(surfaceHolder);
                    }

                    @Override
                    public void surfaceChanged(SurfaceHolder surfaceHolder, int i, int i1, int i2) {

                    }

                    @Override
                    public void surfaceDestroyed(SurfaceHolder surfaceHolder) {
                        cameraSource.stop();
                    }
                });

        barcodeDetector.setProcessor(new Detector.Processor<Barcode>() {
            @Override
            public void release() {

            }

            @Override
            public void receiveDetections(Detector.Detections<Barcode> detections) {
                final SparseArray<Barcode> barCodes = detections.getDetectedItems();
                if (barCodes.size() > 0) {
                    final Barcode barcode = barCodes.valueAt(0);
                    barcodeScanNumber.setText(barcode.displayValue);
                    //TODO something
                }
            }
        });
    }

    @SuppressLint("MissingPermission")
    private void cameraSourceStartFrom(final SurfaceHolder surfaceHolder) {
        try {
            if (!PermissionsUtils.isCameraPermissionGranted(context)) return;
            this.cameraSource.start(surfaceHolder);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/personal/accountantAssistant/ui/barcode/BarcodeScanActivity.kt
package com.personal.accountantAssistant.ui.barcode

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.util.DisplayMetrics
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.widget.EditText
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.vision.CameraSource
import com.google.android.gms.vision.Detector
import com.google.android.gms.vision.Detector.Detections
import com.google.android.gms.vision.barcode.Barcode
import com.google.android.gms.vision.barcode.BarcodeDetector
import com.personal.accountantAssistant.R
import com.personal.accountantAssistant.utils.PermissionsUtils.isCameraPermissionGranted
import java.io.IOException
import java.util.*

class BarcodeScanActivity : AppCompatActivity() {

    private var context: Context? = null
    private var cameraPreviewSurface: SurfaceView? = null
    private var barcodeScanNumber: EditText? = null
    private var cameraSource: CameraSource? = null

    @RequiresApi(Build.VERSION_CODES.R)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_scan_barcode)
        supportActionBar?.setTitle(R.string.barcode_scanner)
        context = this@BarcodeScanActivity
        cameraPreviewSurface = findViewById(R.id.bar_code_scan_view)
        barcodeScanNumber = findViewById(R.id.bar_code_scan_number)
        /*final Button barcodeScanButton = findViewById(R.id.bar_code_scan_button);
        barcodeScanButton.setOnClickListener(view -> finish());*/createCameraSource()
    }

    @RequiresApi(Build.VERSION_CODES.R)
    private fun createCameraSource() {
        val displayMetrics = DisplayMetrics()

        context?.display?.getRealMetrics(displayMetrics)
        val barcodeDetector = BarcodeDetector.Builder(this@BarcodeScanActivity)
                .setBarcodeFormats(Barcode.PRODUCT).build()
        cameraSource = CameraSource.Builder(this@BarcodeScanActivity, barcodeDetector)
                .setFacing(CameraSource.CAMERA_FACING_BACK)
                .setAutoFocusEnabled(true)
                .setRequestedPreviewSize(displayMetrics.widthPixels, displayMetrics.heightPixels)
                .setRequestedFps(24.0f)
                .build()
        cameraPreviewSurface?.holder?.addCallback(object : SurfaceHolder.Callback {
            override fun surfaceCreated(surfaceHolder: SurfaceHolder) {
                cameraSourceStartFrom(surfaceHolder)
            }

            override fun surfaceChanged(surfaceHolder: SurfaceHolder, i: Int, i1: Int, i2: Int) {}
            override fun surfaceDestroyed(surfaceHolder: SurfaceHolder) {
                cameraSource?.stop()
            }
        })
        barcodeDetector.setProcessor(object : Detector.Processor<Barcode> {
            override fun release() {}
            override fun receiveDetections(detections: Detections<Barcode>) {
                val barCodes = detections.detectedItems
                if (barCodes.size() > 0) {
                    val barcode = barCodes.valueAt(0)
                    barcodeScanNumber?.setText(barcode.displayValue)
                    //TODO something
                }
            }
        })
    }

    @SuppressLint("MissingPermission")
    private fun cameraSourceStartFrom(surfaceHolder: SurfaceHolder) {
        try {
            if (!isCameraPermissionGranted(this@BarcodeScanActivity)) return
            cameraSource?.start(surfaceHolder)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}